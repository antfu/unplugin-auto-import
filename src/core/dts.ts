import { ImportsFlatMap } from '../types'

export function generateDeclration(imports: ImportsFlatMap, resolvedImports: ImportsFlatMap = {}) {
  const body = [
    ...Object.entries(imports),
    ...Object.entries(resolvedImports),
  ]
    .map(([name, info]) => `  const ${name}: typeof import('${info.module}')['${info.from || name}']`)
    .join('\n')
  return `// Generated by 'unplugin-auto-import'\ndeclare global {\n${body}\n}\nexport {}\n`
}
