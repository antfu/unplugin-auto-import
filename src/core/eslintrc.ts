import type { ESLintGlobalsPropValue, ESLintrc, ImportsFlatMap } from '../types'

interface ESLintConfigs {
  globals: Record<string, ESLintGlobalsPropValue>
}

// const __comment__ = 'Generated by `unplugin-auto-import`'

export function generateESLintConfigs(
  imports: ImportsFlatMap,
  resolvedImports: ImportsFlatMap = {},
  eslintrc: ESLintrc,
) {
  const eslintConfigs: ESLintConfigs = { globals: {} };

  [
    ...Object.entries(imports),
    ...Object.entries(resolvedImports),
  ]
    .sort((a, b) => a[0].localeCompare(b[0]))
    .forEach(([name]) => { eslintConfigs.globals[name] = eslintrc.globalsPropValue || true })
  const jsonBody = JSON.stringify(eslintConfigs, null, 2)
  return jsonBody
}
