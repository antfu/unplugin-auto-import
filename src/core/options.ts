import { resolve } from 'path'
import { toArray } from '@antfu/utils'
import { createFilter } from '@rollup/pluginutils'
import { isPackageExists } from 'local-pkg'
import { presets } from '../presets'
import { ImportsFlatMap, Options, ResolvedOptions, ResolvedResult } from '../types'

export function resolveOptions(options: Options = {}): ResolvedOptions {
  const imports = flattenImportsMap(options.imports, options.presetOverriding)

  const {
    dts = isPackageExists('typescript'),
  } = options

  const resolved: ResolvedOptions = {
    sourceMap: false,
    resolvedImports: {},
    presetOverriding: false,
    ignore: [],
    ...options,
    dts: dts === false
      ? false
      : dts === true
        ? resolve('auto-imports.d.ts')
        : resolve(dts),
    imports,
    resolvers: toArray(options.resolvers),
    idFilter: createFilter(
      options.include || [/\.[jt]sx?$/, /\.vue$/, /\.vue\?vue/, /\.svelte$/],
      options.exclude || [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/],
    ),
  }

  return resolved
}

export function flattenImportsMap(map: Options['imports'], overriding = false): ImportsFlatMap {
  const flat: ImportsFlatMap = {}
  toArray(map).forEach((definition) => {
    if (typeof definition === 'string') {
      if (!presets[definition])
        throw new Error(`[auto-import] preset ${definition} not found`)
      const preset = presets[definition]
      definition = typeof preset === 'function' ? preset() : preset
    }

    for (const mod of Object.keys(definition)) {
      for (const id of definition[mod]) {
        const meta: ResolvedResult = {
          path: mod,
        }
        let name: string
        if (Array.isArray(id)) {
          name = id[1]
          meta.importName = id[0]
        }
        else {
          name = id
          meta.importName = id
        }

        if (flat[name] && !overriding)
          throw new Error(`[auto-import] identifier ${name} already defined with ${flat[name].path}`)

        flat[name] = meta
      }
    }
  })

  return flat
}
