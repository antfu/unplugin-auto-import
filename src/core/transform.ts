import MagicString from 'magic-string'
import { ImportInfo, TransformOptions, Resolver } from '../types'

const excludeRE = [
  // imported from other module
  /\bimport\s*([\w_$]*?),?\s*\{([\s\S]*?)\}\s*from\b/g,
  // defined as function
  /\bfunction\s*([\s\S]+?)\s*\(/g,
  // defined as local variable
  /\b(?:const|let|var)\s*([\w\d_$]+?)\b/g,
]

const matchRE = /\b(\w+)\b/g
const importAsRE = /^.*\sas\s+/
const multilineCommentsRE = /\/\*(.|[\r\n])*?\*\//gm
const singlelineCommentsRE = /\/\/.*/g

function stripeComments(code: string) {
  return code
    .replace(multilineCommentsRE, '')
    .replace(singlelineCommentsRE, '')
}

export function transform(
  code: string,
  id: string,
  {
    imports,
    sourceMap,
    resolvers,
    resolvedImports = {},
  }: TransformOptions,
) {
  const noComments = stripeComments(code)
  const identifiers = new Set(Array.from(noComments.matchAll(matchRE)).map(i => i[1]))

  // nothing matched, skip
  if (!identifiers.size)
    return null

  // remove those already defined
  for (const regex of excludeRE) {
    Array.from(noComments.matchAll(regex))
      .flatMap(i => [...(i[1]?.split(',') || []), ...(i[2]?.split(',') || [])])
      .map(i => i.replace(importAsRE, '').trim())
      .forEach(i => identifiers.delete(i))
  }

  // nothing matched, skip
  if (!identifiers.size)
    return null

  const modules: Record<string, ImportInfo[]> = {}

  // group by module name
  Array.from(identifiers).forEach((name) => {
    let info: ImportInfo = resolvedImports[name] || imports[name]

    if (!info && resolvers?.length) {
      const resolved = firstNonNullResult(resolvers, name)
      if (resolved) {
        if (typeof resolved === 'string') {
          info = {
            module: resolved,
            name,
            from: 'default',
          }
        }
        else {
          info = resolved
        }
        resolvedImports[name] = info
      }
    }

    if (!info)
      return

    if (!modules[info.module])
      modules[info.module] = []
    modules[info.module].push(info)
  })

  if (!Object.keys(modules).length)
    return

  // stringify import
  const importStatements = Object.entries(modules)
    .map(([moduleName, names]) => {
      const imports = names
        .map(({ name, from }) => from ? `${from} as ${name}` : name)
        .join(', ')
      return `import { ${imports} } from '${moduleName}';`
    })
    .join('')

  const s = new MagicString(code)
  s.prependLeft(0, importStatements)

  return {
    code: s.toString(),
    map: sourceMap ? s.generateMap() : null,
  }
}

function firstNonNullResult(array: Resolver[], name: string) {
  for (let i = 0; i < array.length; i++) {
    const res = array[i](name)
    if (res)
      return res
  }
}
