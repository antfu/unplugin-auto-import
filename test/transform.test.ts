import { resolve } from 'path'
import { promises as fs } from 'fs'
import fg from 'fast-glob'
import { transform } from '../src/core/transform'
import { resolveOptions } from '../src/core/options'

describe('transform', () => {
  const options = resolveOptions({
    imports: [
      'vue',
      'react',
      'svelte',
      'svelte/animate',
      'svelte/easing',
      'svelte/motion',
      'svelte/store',
      'svelte/transition',
      {
        custom: [
          'customNamed',
          ['default', 'customDefault'],
        ],
      },
    ],
  })

  describe('fixtures', () => {
    const root = resolve(__dirname, 'fixtures')
    const files = fg.sync('*', {
      cwd: root,
      onlyFiles: true,
    })

    for (const file of files) {
      it(file, async() => {
        const fixture = await fs.readFile(resolve(root, file), 'utf-8')
        expect(transform(fixture, file, options)?.code ?? fixture).toMatchSnapshot()
      })
    }
  })
})
