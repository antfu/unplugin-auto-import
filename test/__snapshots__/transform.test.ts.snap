// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform fixtures comments.ts 1`] = `
"import { ref, computed } from 'vue';/* import { ref } from 'vue' */
// import { computed } from 'vue'

const foo = ref(0)
const bar = computed(() => {})

// reactive
"
`;

exports[`transform fixtures custom.js 1`] = `
"import { default as customDefault, customNamed } from 'custom';const a = customDefault(customNamed())
"
`;

exports[`transform fixtures non-target.js 1`] = `
"useNonTarget()
"
`;

exports[`transform fixtures react.jsx 1`] = `
"import { useState } from 'react';export function Component() {
  const [count, setCount] = useState()

  return <div>{ count }</div>
}
"
`;

exports[`transform fixtures react-existing.jsx 1`] = `
"import React, { useState } from 'react';

export function Component() {
  const [count, setCount] = useState()

  return <div>{ count }</div>
}
"
`;

exports[`transform fixtures svelte.svelte 1`] = `
"import { writable } from 'svelte/store';import { tweened } from 'svelte/motion';import { cubicOut, quintOut } from 'svelte/easing';import { onMount } from 'svelte';import { flip } from 'svelte/animate';import { fade } from 'svelte/transition';const list = writable([1, 2, 3])
const size = tweened(1, {
  duration: 300,
  easing: cubicOut
});
onMount(() => {})

{#each $list as n (n)}
<div animate:flip=\\"{{delay: 250, duration: 250, easing: quintOut}}\\">
  {n}
</div>
{/each}

{#if condition}
  <div transition:fade=\\"{{amount: 10}}\\">
    fades in and out
  </div>
{/if}
"
`;

exports[`transform fixtures vue.js 1`] = `
"import { ref } from 'vue';const a = ref(0)
"
`;
